docs/developers/source-code.rst

```diff
@@ -0,0 +1,265 @@
+Source code
+===========
+
+Code style
+----------
+
+We intentionally do not have too many code conventions in the Open MPI
+code base.
+
+All languages
+^^^^^^^^^^^^^
+
+* 4 space tabs.  No more, no less.
+* **NEVER** use actual tab characters; always use spaces.  Both emacs
+  and vim have secret mojo that can automatically use spaces when you
+  hit the ``<TAB>`` key.  This makes the code look the same in every
+  browser, regardless of individual tab display settings.
+
+C / C++
+^^^^^^^
+
+* When comparing constants for equality or inequality, always put the
+  constant on the left.  This is defensive programming: if you have a
+  typo in the test and miss a ``!`` or ``=``, you'll get a compiler error.
+  For example:
+
+  .. code-block:: c
+
+     /* Do this */
+     if (NULL == foo) { ... }
+
+     /* Because if you have a typo (i.e., = instead of ==), this will
+        be a compile error rather than a subtle bug */
+     if (NULL = foo) { ... }
+
+* More defensive programming: *always* include blocks in curly braces
+  ``{ }``, even if they are only one line long.  For example:
+
+  .. code-block:: c
+
+     /* Do this */
+     if (whatever) {
+         return OMPI_SUCCESS;
+     }
+
+     /* Not this */
+     if (whatever)
+        return OMPI_SUCCESS;
+
+* Starting with Open MPI 1.7, Open MPI requires a C99-compliant
+  compiler.
+
+  * C++-style comments are now allowed (and preferred).
+  * C99-style mixing declarations are allow allowable (and preferred).
+
+* **ALWAYS** include ``<level>_config.h`` as your first #include file,
+  where ``<level>`` is one of ``ompi``, ``oshmem``, or ``opal`` -- the
+  level that you're writing in.  There are very, very few cases where
+  this is not true (E.g., some bizarre Windows scenarios).  But in
+  99.9999% of cases, this file should be included **first** so that it
+  can affect system-level #include files if necessary.
+* Filenames and symbols must follow the **prefix rule** (see [e-mail
+  thread](http://www.open-mpi.org/community/lists/devel/2009/07/6389.php)):
+
+  * Filenames must be prefixed with ``<framework>_<component>``.
+  * Public symbols must be prefixed in components with
+    ``<layer>_<framework>_<component>``, where ``<layer>`` is one of
+    ``mca``, ``ompi``, ``oshmem``, or ``opal``.  Note that `mca` used
+    to be the most common, but it has fallen out of favor compared to
+    the other ``<layer>`` prefixes.  When in doubt about whether a
```

Comment:
Should we use `<project>` here instead of `<layer>` since we refer to opal/ompi as projects later in this document?